tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -353,22 +353,22 @@ var _ = ginkgo.Describe("[tidb-operator] TiDBCluster", func() {
 		svcType := corev1.ServiceTypeNodePort
 		trafficPolicy := corev1.ServiceExternalTrafficPolicyTypeLocal
 
-		err = wait.PollImmediate(5*time.Second, 5*time.Minute, func() (bool, error) {
-			tc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
-			framework.ExpectNoError(err, "Expected get TiDB cluster")
+		tc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
+		framework.ExpectNoError(err, "Expected get TiDB cluster")
+
+		ginkgo.By("Updating TiDB service")
+		err = controller.GuaranteedUpdate(genericCli, tc, func() error {
 			tc.Spec.TiDB.Service.Type = svcType
 			tc.Spec.TiDB.Service.ExternalTrafficPolicy = &trafficPolicy
 			tc.Spec.TiDB.Service.Annotations = map[string]string{
 				"test": "test",
 			}
-			_, err = cli.PingcapV1alpha1().TidbClusters(ns).Update(tc)
```

Comment:
The original code is ok because it will retry on conflict error in wait retries. However, with `controller.GuaranteedUpdate` , the code can be simplified and we can split the whole process into two steps:

- Update the CRD (retry on conflict)
- Wait and verify